crypto/x509v3/v3_purp.c

```diff
@@ -471,8 +463,7 @@ static void x509v3_cache_extensions(X509 *x)
     if (!X509_NAME_cmp(X509_get_subject_name(x), X509_get_issuer_name(x))) {
         x->ex_flags |= EXFLAG_SI;
         /* If SKID matches AKID also indicate self signed */
-        if (X509_check_akid(x, x->akid) == X509_V_OK &&
-            !ku_reject(x, KU_KEY_CERT_SIGN))
+        if (X509_check_akid(x, x->akid) == X509_V_OK)
```

Comment:
how about this:
```
--- a/crypto/x509v3/v3_purp.c
+++ b/crypto/x509v3/v3_purp.c
@@ -463,8 +463,16 @@ static void x509v3_cache_extensions(X509 *x)
     if (!X509_NAME_cmp(X509_get_subject_name(x), X509_get_issuer_name(x))) {
         x->ex_flags |= EXFLAG_SI;
         /* If SKID matches AKID also indicate self signed */
-        if (X509_check_akid(x, x->akid) == X509_V_OK)
-            x->ex_flags |= EXFLAG_SS;
+        if (X509_check_akid(x, x->akid) == X509_V_OK) {
+            X509_PUBKEY *xpkey = X509_get_X509_PUBKEY(x);
+            if (xpkey != NULL) {
+                EVP_PKEY *pkey = X509_PUBKEY_get0(xpkey);
+                if (pkey != NULL
+                        && EVP_PKEY_id(pkey) != EVP_PKEY_X25519
+                        && EVP_PKEY_id(pkey) != EVP_PKEY_X448)
+                    x->ex_flags |= EXFLAG_SS;
+            }
+        }
     }
     x->altname = X509_get_ext_d2i(x, NID_subject_alt_name, NULL, NULL);
     x->nc = X509_get_ext_d2i(x, NID_name_constraints, &i, NULL);
```

It can't be self-signed when we are unable to verify the signature using the public key.

Reply:
Adding references to specific crypto algorithms would add quite some clutter to the verification functions, which are already very hard to follow. If done for these two algorithms, then consequently putting in many more such exceptions etc. would be in order. 

The RFC 5280 does not do this, and I'd recommend not to do this in the OpenSSL implementation, either.